File: leetcode/problems/language-specific/go/3.longest-substring-without-repeating-characters.go
/*
 * @lc app=leetcode id=3 lang=golang
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	var exists = struct{}{}
	res, hashSet, left, right := 0, make(map[byte]struct{}), 0, 0
	for right < len(s) {
		if _, ok := hashSet[s[right]]; !ok {
			hashSet[s[right]] = exists
			res = findMax(res, len(hashSet))
			right++
		} else {
			delete(hashSet, s[left])
			left++
		}
	}

	return res
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

// @lc code=end

File: leetcode/problems/language-specific/python/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start

# * Two Pointers and Sliding Window Solution | O(n) Time | O(n) Space
# * n -> The length of s string

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        s_set = set()
        longest_substring_length = 0
        left, right = 0, 0
        while right < len(s):
            if s[right] not in s_set:
                s_set.add(s[right])
                longest_substring_length = max(
                    longest_substring_length, len(s_set))
                right += 1

            else:
                s_set.remove(s[left])
                left += 1

        return longest_substring_length

# @lc code=end
